// pivotal-ui/react/expander/expander.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports.ExpanderContent = exports.ExpanderTrigger = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _collapsible = require("../collapsible");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var ExpanderTrigger = /*#__PURE__*/function (_React$PureComponent) {
  _inheritsLoose(ExpanderTrigger, _React$PureComponent);

  function ExpanderTrigger(props, context) {
    var _this;

    _this = _React$PureComponent.call(this, props, context) || this;

    _defineProperty(_assertThisInitialized(_this), "setTarget", function (target) {
      return _this.setState({
        target: target
      });
    });

    _defineProperty(_assertThisInitialized(_this), "toggleExpander", function (event) {
      event.preventDefault();

      if (_this.state.target) {
        _this.state.target.toggle();
      } else {
        console.warn('No ExpanderContent provided to ExpanderTrigger.');
      }
    });

    _this.state = {};
    return _this;
  }

  var _proto = ExpanderTrigger.prototype;

  _proto.render = function render() {
    return _react["default"].cloneElement(this.props.children, {
      onClick: this.toggleExpander
    });
  };

  return ExpanderTrigger;
}(_react["default"].PureComponent);

exports.ExpanderTrigger = ExpanderTrigger;

var ExpanderContent = /*#__PURE__*/function (_React$PureComponent2) {
  _inheritsLoose(ExpanderContent, _React$PureComponent2);

  function ExpanderContent(props, context) {
    var _this2;

    _this2 = _React$PureComponent2.call(this, props, context) || this;

    _defineProperty(_assertThisInitialized(_this2), "toggle", function () {
      return _this2.setState({
        expanded: !_this2.state.expanded
      });
    });

    _this2.state = {
      expanded: _this2.props.expanded
    };
    return _this2;
  }

  var _proto2 = ExpanderContent.prototype;

  _proto2.render = function render() {
    var expanded = this.state.expanded;
    return _react["default"].createElement(_collapsible.Collapsible, _extends({}, this.props, {
      expanded: expanded
    }));
  };

  return ExpanderContent;
}(_react["default"].PureComponent);

exports.ExpanderContent = ExpanderContent;

_defineProperty(ExpanderContent, "propTypes", {
  expanded: _propTypes["default"].bool
});
