// pivotal-ui/react/panes/panes.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports.Pane = exports.BasePane = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _helpers = require("../helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var BasePane = /*#__PURE__*/function (_React$PureComponent) {
  _inheritsLoose(BasePane, _React$PureComponent);

  function BasePane() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = BasePane.prototype;

  _proto.componentDidMount = function componentDidMount() {
    require('../../css/panes');
  };

  _proto.render = function render() {
    var _this$props = this.props,
        innerClassName = _this$props.innerClassName,
        children = _this$props.children,
        other = _objectWithoutPropertiesLoose(_this$props, ["innerClassName", "children"]);

    var outerProps = (0, _helpers.mergeProps)(other, {
      className: 'pane'
    });
    var innerProps = (0, _helpers.mergeProps)({
      className: innerClassName
    }, {
      className: 'container'
    });
    return _react["default"].createElement("div", outerProps, _react["default"].createElement("div", innerProps, children));
  };

  return BasePane;
}(_react["default"].PureComponent);

exports.BasePane = BasePane;

_defineProperty(BasePane, "propTypes", {
  className: _propTypes["default"].string,
  innerClassName: _propTypes["default"].string
});

var Pane = /*#__PURE__*/function (_React$PureComponent2) {
  _inheritsLoose(Pane, _React$PureComponent2);

  function Pane() {
    return _React$PureComponent2.apply(this, arguments) || this;
  }

  var _proto2 = Pane.prototype;

  _proto2.componentDidMount = function componentDidMount() {
    require('../../css/panes');
  };

  _proto2.render = function render() {
    var _this$props2 = this.props,
        className = _this$props2.className,
        other = _objectWithoutPropertiesLoose(_this$props2, ["className"]);

    return _react["default"].createElement(BasePane, _extends({}, other, {
      className: className
    }));
  };

  return Pane;
}(_react["default"].PureComponent);

exports.Pane = Pane;

_defineProperty(Pane, "propTypes", {
  className: _propTypes["default"].string
});
