// pivotal-ui/react/siteframe/sidebar.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports.Sidebar = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _sidebar_links = require("./sidebar_links");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Sidebar = /*#__PURE__*/function (_React$PureComponent) {
  _inheritsLoose(Sidebar, _React$PureComponent);

  function Sidebar() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = Sidebar.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        className = _this$props.className,
        primaryLinks = _this$props.primaryLinks,
        secondaryLinks = _this$props.secondaryLinks,
        renderLink = _this$props.renderLink;
    return _react["default"].createElement("nav", {
      className: (0, _classnames["default"])('pui-siteframe-sidebar', className)
    }, _react["default"].createElement(_sidebar_links.SidebarLinks, {
      links: primaryLinks,
      renderLink: renderLink,
      className: 'pui-sidebar-primary-links'
    }), secondaryLinks && _react["default"].createElement(_sidebar_links.SidebarLinks, {
      links: secondaryLinks,
      renderLink: renderLink,
      className: 'pui-sidebar-secondary-links'
    }));
  };

  return Sidebar;
}(_react["default"].PureComponent);

exports.Sidebar = Sidebar;

_defineProperty(Sidebar, "propTypes", {
  className: _propTypes["default"].string,
  primaryLinks: _propTypes["default"].array.isRequired,
  secondaryLinks: _propTypes["default"].array,
  renderLink: _propTypes["default"].func.isRequired
});
