// pivotal-ui/react/siteframe/siteframe.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports.Siteframe = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _header = require("./header");

var _sidebar = require("./sidebar");

var _flexGrids = require("../flex-grids");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Siteframe = /*#__PURE__*/function (_React$PureComponent) {
  _inheritsLoose(Siteframe, _React$PureComponent);

  function Siteframe() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = Siteframe.prototype;

  _proto.componentDidMount = function componentDidMount() {
    require('../../css/siteframe');
  };

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        headerProps = _this$props.headerProps,
        sidebarProps = _this$props.sidebarProps;
    return _react["default"].createElement("div", {
      className: "pui-siteframe"
    }, headerProps && _react["default"].createElement(_header.Header, headerProps), _react["default"].createElement(_flexGrids.Grid, {
      gutter: false,
      className: "pui-siteframe-body"
    }, sidebarProps && _react["default"].createElement(_flexGrids.FlexCol, {
      fixed: true
    }, _react["default"].createElement(_sidebar.Sidebar, sidebarProps)), _react["default"].createElement(_flexGrids.FlexCol, null, children)));
  };

  return Siteframe;
}(_react["default"].PureComponent);

exports.Siteframe = Siteframe;

_defineProperty(Siteframe, "propTypes", {
  children: _propTypes["default"].node,
  headerProps: _propTypes["default"].object,
  sidebarProps: _propTypes["default"].object
});
