// pivotal-ui/react/radio/radio_group.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports.RadioGroup = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var RadioGroup = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(RadioGroup, _React$Component);

  function RadioGroup() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = RadioGroup.prototype;

  _proto.componentDidMount = function componentDidMount() {
    require('../../css/forms');
  };

  _proto.render = function render() {
    var _this$props = this.props,
        name = _this$props.name,
        children = _this$props.children,
        onChange = _this$props.onChange,
        className = _this$props.className,
        value = _this$props.value,
        others = _objectWithoutPropertiesLoose(_this$props, ["name", "children", "onChange", "className", "value"]);

    var radioProps = onChange ? {
      name: name,
      onChange: onChange
    } : {
      name: name,
      readOnly: true
    };
    return _react["default"].createElement("div", _extends({}, others, {
      className: (0, _classnames["default"])('pui-radio-group', className)
    }), _react["default"].Children.map(children, function (child) {
      return _react["default"].cloneElement(child, _extends({}, radioProps, {
        checked: child.props.value === value
      }));
    }));
  };

  return RadioGroup;
}(_react["default"].Component);

exports.RadioGroup = RadioGroup;

_defineProperty(RadioGroup, "propTypes", {
  id: _propTypes["default"].string,
  name: _propTypes["default"].string,
  onChange: _propTypes["default"].func,
  value: _propTypes["default"].any
});
