// pivotal-ui/react/radio/radio.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports.Radio = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _lodash = _interopRequireDefault(require("lodash.uniqueid"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Radio = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Radio, _React$Component);

  function Radio() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = Radio.prototype;

  _proto.componentDidMount = function componentDidMount() {
    require('../../css/forms');

    require('../../css/radio');
  };

  _proto.render = function render() {
    var _this$props = this.props,
        className = _this$props.className,
        disabled = _this$props.disabled,
        children = _this$props.children,
        labelClassName = _this$props.labelClassName,
        noSelect = _this$props.noSelect,
        style = _this$props.style,
        _this$props$id = _this$props.id,
        id = _this$props$id === void 0 ? (0, _lodash["default"])('radio') : _this$props$id,
        others = _objectWithoutPropertiesLoose(_this$props, ["className", "disabled", "children", "labelClassName", "noSelect", "style", "id"]);

    return _react["default"].createElement("div", {
      className: (0, _classnames["default"])('pui-radio', className),
      style: style
    }, _react["default"].createElement("input", _extends({}, others, {
      className: 'pui-radio-input',
      type: 'radio',
      id: id,
      disabled: disabled,
      'aria-disabled': disabled
    })), _react["default"].createElement("label", {
      className: (0, _classnames["default"])('pui-radio-label', labelClassName, {
        'pui-no-select': noSelect
      }),
      htmlFor: id
    }, _react["default"].createElement("span", {
      className: "pui-radio-circle"
    }), children));
  };

  return Radio;
}(_react["default"].Component);

exports.Radio = Radio;

_defineProperty(Radio, "propTypes", {
  checked: _propTypes["default"].bool,
  className: _propTypes["default"].string,
  defaultChecked: _propTypes["default"].bool,
  disabled: _propTypes["default"].bool,
  id: _propTypes["default"].string,
  labelClassName: _propTypes["default"].string,
  name: _propTypes["default"].string,
  noSelect: _propTypes["default"].bool,
  onChange: _propTypes["default"].func,
  style: _propTypes["default"].object,
  value: _propTypes["default"].string.isRequired
});
