// pivotal-ui/react/media/media.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports.Flag = exports.Media = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames2 = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var shortSizes = {
  xsmall: 'xs',
  small: 'sm',
  medium: 'md',
  large: 'lg'
};
var charSizes = {
  small: 's',
  medium: 'm',
  large: 'l',
  xlarge: 'xl'
};
var paddingDirection = {
  left: 'r',
  right: 'l'
};

var Media = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Media, _React$Component);

  function Media() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = Media.prototype;

  _proto.componentDidMount = function componentDidMount() {
    require('../../css/media');
  };

  _proto.render = function render() {
    var _classnames;

    var _this$props = this.props,
        className = _this$props.className,
        innerClassName = _this$props.innerClassName,
        image = _this$props.image,
        mediaSpacing = _this$props.mediaSpacing,
        stackSize = _this$props.stackSize,
        vAlign = _this$props.vAlign,
        placement = _this$props.placement,
        children = _this$props.children,
        other = _objectWithoutPropertiesLoose(_this$props, ["className", "innerClassName", "image", "mediaSpacing", "stackSize", "vAlign", "placement", "children"]);

    var vAlignClass = vAlign && "media-" + vAlign;
    var classes = (0, _classnames2["default"])('media', stackSize && "media-stackable-" + shortSizes[stackSize], className);
    var bodyClasses = (0, _classnames2["default"])('media-body', vAlignClass, innerClassName);
    var mediaClasses = (0, _classnames2["default"])("media-" + placement, vAlignClass, (_classnames = {}, _classnames["p" + paddingDirection[placement] + charSizes[mediaSpacing]] = charSizes[mediaSpacing], _classnames));
    var content = [_react["default"].createElement("div", {
      key: 0,
      className: mediaClasses
    }, image), _react["default"].createElement("div", {
      key: 1,
      className: bodyClasses
    }, children)];
    if (placement === 'right') content.reverse();
    return _react["default"].createElement("div", _extends({}, other, {
      className: classes
    }), content);
  };

  return Media;
}(_react["default"].Component);

exports.Media = Media;

_defineProperty(Media, "propTypes", {
  image: _propTypes["default"].oneOfType([_propTypes["default"].node, _propTypes["default"].object]).isRequired,
  innerClassName: _propTypes["default"].string,
  mediaSpacing: _propTypes["default"].oneOf(['small', 'medium', 'large', 'xlarge']),
  stackSize: _propTypes["default"].oneOf(['xsmall', 'small', 'medium', 'large']),
  vAlign: _propTypes["default"].oneOf(['middle', 'bottom', 'top']),
  placement: _propTypes["default"].oneOf(['left', 'right']),
  className: _propTypes["default"].string
});

_defineProperty(Media, "defaultProps", {
  placement: 'left'
});

var Flag = /*#__PURE__*/function (_React$Component2) {
  _inheritsLoose(Flag, _React$Component2);

  function Flag() {
    return _React$Component2.apply(this, arguments) || this;
  }

  var _proto2 = Flag.prototype;

  _proto2.componentDidMount = function componentDidMount() {
    require('../../css/media');
  };

  _proto2.render = function render() {
    return _react["default"].createElement(Media, _extends({}, this.props, {
      vAlign: "middle"
    }));
  };

  return Flag;
}(_react["default"].Component);

exports.Flag = Flag;
