// pivotal-ui/react/tabs/tab.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports.Tab = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Tab = /*#__PURE__*/function (_React$PureComponent) {
  _inheritsLoose(Tab, _React$PureComponent);

  function Tab() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = Tab.prototype;

  _proto.render = function render() {
    return null;
  };

  return Tab;
}(_react["default"].PureComponent);

exports.Tab = Tab;

_defineProperty(Tab, "propTypes", {
  'aria-labelledby': _propTypes["default"].string,
  className: _propTypes["default"].string,
  disabled: _propTypes["default"].bool,
  eventKey: _propTypes["default"].any,
  onEntered: _propTypes["default"].func,
  onExited: _propTypes["default"].func,
  tabClassName: _propTypes["default"].string,
  title: _propTypes["default"].node.isRequired
});

_defineProperty(Tab, "defaultProps", {
  onEntered: function onEntered() {},
  onExited: function onExited() {}
});
