// pivotal-ui/react/tabs/tab_content.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _classnames2 = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var TabContent = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(TabContent, _React$Component);

  function TabContent() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = TabContent.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        activeKey = _this$props.activeKey,
        childArray = _this$props.childArray,
        id = _this$props.id,
        isLeft = _this$props.isLeft,
        paneWidth = _this$props.paneWidth,
        transitionProgress = _this$props.transitionProgress;
    var tabContent = null;
    var leftPaneClasses = "col-xs-" + paneWidth;
    childArray.forEach(function (child, key) {
      var _classnames;

      var _child$props = child.props,
          ariaLabelledBy = _child$props['aria-labelledby'],
          children = _child$props.children,
          className = _child$props.className,
          eventKey = _child$props.eventKey,
          onEntered = _child$props.onEntered,
          onExited = _child$props.onExited,
          tabClassName = _child$props.tabClassName,
          title = _child$props.title,
          props = _objectWithoutPropertiesLoose(_child$props, ["aria-labelledby", "children", "className", "eventKey", "onEntered", "onExited", "tabClassName", "title"]);

      var paneId = id + "-pane-" + key;
      var tabId = id + "-tab-" + key;
      var isActive = eventKey === activeKey;
      var style = transitionProgress < 1 ? {
        opacity: Math.abs(1 - 2 * transitionProgress)
      } : {};
      if (!isActive) return false;
      tabContent = _react["default"].createElement("div", _extends({
        className: (0, _classnames2["default"])('tab-content', (_classnames = {}, _classnames[leftPaneClasses] = isLeft, _classnames), className)
      }, props), _react["default"].createElement("div", {
        className: "tab-pane fade active in",
        id: paneId,
        role: "tabpanel",
        "aria-labelledby": ariaLabelledBy || tabId,
        "aria-hidden": "false",
        style: style
      }, children));
    });
    return tabContent;
  };

  return TabContent;
}(_react["default"].Component);

exports["default"] = TabContent;

_defineProperty(TabContent, "propTypes", {
  activeKey: _propTypes["default"].any,
  childArray: _propTypes["default"].array,
  id: _propTypes["default"].string,
  isLeft: _propTypes["default"].bool,
  paneWidth: _propTypes["default"].number,
  transitionProgress: _propTypes["default"].number
});
