// pivotal-ui/react/progress-bar/progress_bar.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports.ProgressBar = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var ProgressBar = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(ProgressBar, _React$Component);

  function ProgressBar() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = ProgressBar.prototype;

  _proto.componentDidMount = function componentDidMount() {
    require('../../css/progress-bar');
  };

  _proto.render = function render() {
    var _this$props = this.props,
        barClassName = _this$props.barClassName,
        className = _this$props.className,
        value = _this$props.value;
    return _react["default"].createElement("div", {
      className: (0, _classnames["default"])('pui-progress', className)
    }, _react["default"].createElement("div", {
      className: (0, _classnames["default"])('pui-progress-bar', barClassName),
      role: 'progressbar',
      'aria-valuemax': 100,
      'aria-valuemin': 0,
      'aria-valuenow': value,
      style: {
        width: value + "%"
      }
    }));
  };

  return ProgressBar;
}(_react["default"].Component);

exports.ProgressBar = ProgressBar;

_defineProperty(ProgressBar, "propTypes", {
  barClassName: _propTypes["default"].string,
  value: _propTypes["default"].number.isRequired
});

_defineProperty(ProgressBar, "defaultProps", {
  value: 0
});
