// pivotal-ui/react/collapse/collapse.js
/* (c) Copyright 2020 Pivotal Software, Inc. All Rights Reserved. */
"use strict";

exports.__esModule = true;
exports.AltCollapse = exports.Collapse = exports.BaseCollapse = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _collapsible = require("../collapsible");

var _helpers = require("../helpers");

var _iconography = require("../iconography");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var BaseCollapse = /*#__PURE__*/function (_React$PureComponent) {
  _inheritsLoose(BaseCollapse, _React$PureComponent);

  function BaseCollapse(props, context) {
    var _this;

    _this = _React$PureComponent.call(this, props, context) || this;

    _defineProperty(_assertThisInitialized(_this), "handleSelect", function (e) {
      e.preventDefault();

      _this.setState({
        expanded: !_this.state.expanded
      });
    });

    _this.state = {
      expanded: !!props.defaultExpanded
    };
    return _this;
  }

  var _proto = BaseCollapse.prototype;

  _proto.componentDidMount = function componentDidMount() {
    require('../../css/collapse');
  };

  _proto.renderHeader = function renderHeader() {
    var header = this.props.header;
    var expanded = this.state.expanded;
    return _react["default"].createElement("a", {
      href: "#",
      "aria-expanded": expanded,
      "aria-selected": expanded
    }, header);
  };

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        defaultExpanded = _this$props.defaultExpanded,
        divider = _this$props.divider,
        header = _this$props.header,
        others = _objectWithoutPropertiesLoose(_this$props, ["children", "defaultExpanded", "divider", "header"]);

    var props = (0, _helpers.mergeProps)(others, {
      className: ['pui-collapse', {
        'pui-collapse-divider': divider
      }]
    });
    var expanded = this.state.expanded;
    return _react["default"].createElement("div", props, _react["default"].createElement("div", {
      className: "pui-collapse-panel-heading",
      onClick: this.handleSelect
    }, _react["default"].createElement("div", {
      className: "pui-collapse-panel-title",
      role: "presentation"
    }, this.renderHeader())), _react["default"].createElement("div", {
      className: "pui-collapse-panel"
    }, _react["default"].createElement(_collapsible.Collapsible, {
      className: "pui-collapse-panel-body",
      expanded: expanded,
      delay: 200
    }, children)));
  };

  return BaseCollapse;
}(_react["default"].PureComponent);

exports.BaseCollapse = BaseCollapse;

_defineProperty(BaseCollapse, "propTypes", {
  divider: _propTypes["default"].bool,
  header: _propTypes["default"].node.isRequired,
  defaultExpanded: _propTypes["default"].bool
});

var Collapse = /*#__PURE__*/function (_BaseCollapse) {
  _inheritsLoose(Collapse, _BaseCollapse);

  function Collapse() {
    return _BaseCollapse.apply(this, arguments) || this;
  }

  var _proto2 = Collapse.prototype;

  _proto2.renderHeader = function renderHeader() {
    var header = this.props.header;
    var expanded = this.state.expanded;
    var iconSrc = expanded ? 'arrow_drop_down' : 'arrow_drop_right';
    return _react["default"].createElement("div", {
      className: "pui-collapse-trigger collapse-trigger"
    }, _react["default"].createElement(_iconography.Icon, {
      className: "pui-collapse-icon collapse-icon",
      src: iconSrc
    }), header);
  };

  return Collapse;
}(BaseCollapse);

exports.Collapse = Collapse;

var AltCollapse = /*#__PURE__*/function (_BaseCollapse2) {
  _inheritsLoose(AltCollapse, _BaseCollapse2);

  function AltCollapse() {
    return _BaseCollapse2.apply(this, arguments) || this;
  }

  var _proto3 = AltCollapse.prototype;

  _proto3.renderHeader = function renderHeader() {
    var header = this.props.header;
    var expanded = this.state.expanded;
    var iconSrc = expanded ? 'remove_circle' : 'add_circle';
    return _react["default"].createElement("div", {
      className: "pui-collapse-trigger collapse-trigger"
    }, _react["default"].createElement(_iconography.Icon, {
      className: "pui-collapse-icon collapse-icon",
      src: iconSrc
    }), header);
  };

  return AltCollapse;
}(BaseCollapse);

exports.AltCollapse = AltCollapse;
